### PR TITLE
Add new primitive: Frustum

### PR DESCRIPTION
**Pull request by cserteGT3**
I started to implement one of my long time-planned ideas: a truncated cone. Before diving deeper into the coding, I wanted to ask if something like that would be a good addition to the list of primitives? I was also thinking that maybe a more general implementation (handling polygonal bottom and top) would be better, but that is way above my knowledge (and my usecase as well).

Regarding the implementation:
- I can roughly copy the list of needed methods from other primitives, but not sure if are there any really/not so important functions.
- Later I am planning to take on the visualization of both Cone and TruncatedCone, but that will be the topic of another PR.
- I am happy to work on this, but don't want to burden you with every single detail, so I'm not sure how often to ask for feedback. But I assume it will turn out ok :)

### CODE REVIEW

**Comment by juliohm** [comment]
That is great @cserteGT3 , thanks for considering a new primitive. What kinds of applications you have in mind with the truncated cone?

**Comment by cserteGT3** [comment]
> That is great @cserteGT3 , thanks for considering a new primitive. What kinds of applications you have in mind with the truncated cone?

I am doing some basic point cloud processing and feature recognition and would be more practical to represent conical surfaces on these parts with truncated cones instead of "infinite" ones.  I would/will code this anyways and though that instead of doing it quick and dirty for myself, it's a great oppurtinity to practice coding and better understanding the Meshes package. And I also enjoy giving back to the julia community :)

**Comment by juliohm** [comment]
Looking forward to the contributions! Feel free to ping me here anytime, or send a message over our Zulip if you need to clarify something with a quick chat.

**Comment by cserteGT3** [comment]
As far as can I see, I have already implemented most of the methods for the truncated cone (except the visualizatio/discretization stuff). Though I would have a few questions:
* I'm copying mostly the cylinder, cone and cylindersurface implementation. I don't get the purpose of the `CylinderSurface` and `ConeSurface` types. Should `TruncatedCone` has one as well? (I feel like it should not, but not sure why.)
* `Cylinder`, `CylinderSurface`, `Cone` and `ConeSurface` don't have implementation for `boundingbox`. Is there a design decision why those are missing, or they just weren't in focus? I assume that truncated cone should have `boundingbox`.
* I am not sure about the `paramdim` function. I don't think that the truncated cone can be parameterized, so it should be `nothing`?

Sorry for the abstract questions :)

**Comment by juliohm** [comment]
> * I don't get the purpose of the `CylinderSurface` and `ConeSurface` types. Should `TruncatedCone` has one as well?

The primitives with the `Surface` suffix represent the boundary of the other primitive without the suffix. So for example, the `Cylinder` is a 3D volume. Its boundary is the `CylinderSurface` with 2D area.

I think we should consider using the name Frustum and FrustumSurface for the truncated cone. What do you think?

> * `Cylinder`, `CylinderSurface`, `Cone` and `ConeSurface` don't have implementation for `boundingbox`. Is there a design decision why those are missing, or they just weren't in focus? I assume that truncated cone should have `boundingbox`.

They do have `boundingbox`. The fallback is implemented for all Primitive types.

> * I am not sure about the `paramdim` function. I don't think that the truncated cone can be parameterized, so it should be `nothing`?

The parametric dimension tells the "true" dimension of the objects regardless of where it is embedded. If the Frustum is a volume, then paramdim should be 3. If it is not paramterized, that is no problem, we have the `isparametrized` trait for that, which defaults to `false`.

**Comment by cserteGT3** [comment]
> The primitives with the `Surface` suffix represent the boundary of the other primitive without the suffix. So for example, the `Cylinder` is a 3D volume. Its boundary is the `CylinderSurface` with 2D area.
> 
> I think we should consider using the name Frustum and FrustumSurface for the truncated cone. What do you think?

Thanks! I agree that Frustum and FrustumSurface are better, updated the code.

> > * `Cylinder`, `CylinderSurface`, `Cone` and `ConeSurface` don't have implementation for `boundingbox`. Is there a design decision why those are missing, or they just weren't in focus? I assume that truncated cone should have `boundingbox`.
> 
> They do have `boundingbox`. The fallback is implemented for all Primitive types.

Yeas but the fallback doesn't work, as it falls back to `boundingbox(nothing)` in the end, if I understand correctly. Should I open an issue for that?

```julia
julia> p = Point3(0,0,0)
Point(0.0, 0.0, 0.0)

julia> a = Point3(9,0,0)
Point(9.0, 0.0, 0.0)

julia> c = Cylinder(p, a, 15)
Cylinder{3,Float64}
├─ bot: Plane(p: (0.0, 0.0, 0.0), u: (-0.0, 1.0, -0.0), v: (-0.0, -0.0, 1.0))
├─ top: Plane(p: (9.0, 0.0, 0.0), u: (-0.0, 1.0, -0.0), v: (-0.0, -0.0, 1.0))
└─ radius: 15.0

julia> boundingbox(c)
ERROR: MethodError: no method matching iterate(::Nothing)

Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen})
   @ Base range.jl:880
  iterate(::Union{LinRange, StepRangeLen}, ::Integer)
   @ Base range.jl:880
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}}
   @ Base dict.jl:698
  ...

Stacktrace:
  [1] iterate(::Base.Generator{Nothing, Meshes.var"#281#282"})
    @ Base .\generator.jl:44
  [2] iterate(::Base.Generator{Base.Generator{Nothing, Meshes.var"#281#282"}, Meshes.var"#285#286"})
    @ Base .\generator.jl:44
  [3] iterate(f::Base.Iterators.Flatten{Base.Generator{Base.Generator{Nothing, Meshes.var"#281#282"}, Meshes.var"#285#286"}}, state::Tuple{})
    @ Base.Iterators .\iterators.jl:1195
  [4] iterate(f::Base.Iterators.Flatten{Base.Generator{Base.Generator{Nothing, Meshes.var"#281#282"}, Meshes.var"#285#286"}})
    @ Base.Iterators .\iterators.jl:1191
  [5] first(itr::Base.Iterators.Flatten{Base.Generator{Base.Generator{Nothing, Meshes.var"#281#282"}, Meshes.var"#285#286"}})
    @ Base .\abstractarray.jl:465
  [6] _pboxes(points::Base.Iterators.Flatten{Base.Generator{Base.Generator{Nothing, Meshes.var"#281#282"}, Meshes.var"#285#286"}})
    @ Meshes C:\Users\csert\.julia\dev\Meshes\src\boundingboxes.jl:52
  [7] _bboxes(boxes::Base.Generator{Nothing, Meshes.var"#281#282"})
    @ Meshes C:\Users\csert\.julia\dev\Meshes\src\boundingboxes.jl:49
  [8] boundingbox(geoms::Nothing)
    @ Meshes C:\Users\csert\.julia\dev\Meshes\src\boundingboxes.jl:22
  [9] boundingbox(p::CylinderSurface{Float64})
    @ Meshes C:\Users\csert\.julia\dev\Meshes\src\boundingboxes.jl:18
 [10] boundingbox(p::Cylinder{Float64})
    @ Meshes C:\Users\csert\.julia\dev\Meshes\src\boundingboxes.jl:18
 [11] top-level scope
    @ REPL[9]:1

```

> The parametric dimension tells the "true" dimension of the objects regardless of where it is embedded. If the Frustum is a volume, then paramdim should be 3. If it is not paramterized, that is no problem, we have the `isparametrized` trait for that, which defaults to `false`.

I think it is not parameterized, so I did not touch that part.

What do you think of the current state of the implementation?

**Comment by juliohm** [comment]
> Yeas but the fallback doesn't work, as it falls back to `boundingbox(nothing)` in the end, if I understand correctly. Should I open an issue for that?

That would be great. The problem is that we didn't implement the boundingbox for the CylinderSurface in this case. We just need to add methods for the *Surface types because the other primitives will fallback to the bounding box of the boundary.

File: src/Meshes.jl [hunk 1]
```diff
@@ -146,6 +146,8 @@ export
   Cone,
   ConeSurface,
   Torus,
+  Frustum,
+  FrustumSurface,
```
Comment (juliohm) [inline src/Meshes.jl:149]: Let's move them to before the Torus

File: src/predicates/in.jl [hunk 2]
```diff
@@ -90,6 +90,24 @@ function Base.in(p::Point{3,T}, t::Torus{T}) where {T}
...
+  ax_d = (p - t) ⋅ normalize(ax)
+  ax_drel = ax_d/norm(ax)
+  # cone radius at axial distance of p
+  rt = radius(top(f))
+  rb = radius(bottom(f))
+  c_r = rt*(1-ax_drel) + rb*ax_drel
+  # radial distance of p
+  rad_d = norm((p - t) - ax_drel * ax)
+  rad_d ≤ c_r
+end
```
Comment (juliohm) [inline src/predicates/in.jl]: Same here, before the Torus

File: src/predicates/isconvex.jl [hunk 3]
```diff
@@ -52,6 +52,8 @@ isconvex(::CylinderSurface) = false
 
 isconvex(::Torus) = false
 
+isconvex(::Frustum) = true
```
Comment (juliohm) [inline src/predicates/isconvex.jl:53]: before Torus

File: src/primitives.jl [hunk 4]
```diff
@@ -52,3 +52,5 @@ include("primitives/cylindersurface.jl")
 include("primitives/cone.jl")
 include("primitives/conesurface.jl")
 include("primitives/torus.jl")
+include("primitives/frustum.jl")
+include("primitives/frustumsurface.jl")
```
Comment (juliohm) [inline src/primitives.jl:55]: before Torus

File: test/primitives.jl [hunk 5]
```diff
@@ -1114,3 +1114,65 @@
...
+
+@testset "Frustum" begin
+  pb = Plane(P3(0, 0, 0), V3(0, 0, 1))
+  db = Disk(pb, T(1))
+  pt = Plane(P3(0, 0, 10), V3(0, 0, 1))
+  dt = Disk(pt, T(2))
+  f = Frustum(db, dt)
+  @test embeddim(f) == 3
+  @test coordtype(f) == T
+  @test boundary(f) == FrustumSurface(db, dt)
```
Comment (juliohm) [inline test/primitives.jl]: before Torus

**Comment by juliohm** [comment]
@cserteGT3 can you also confirm that the difference between the current Frustum implementation and the Cylinder is that the bottom and top disks can have different radius? Do you think it would be too much more work to support polygonal bases?

**Comment by cserteGT3** [comment]
> @cserteGT3 can you also confirm that the difference between the current Frustum implementation and the Cylinder is that the bottom and top disks can have different radius?

Yes.

> Do you think it would be too much more work to support polygonal bases?

When started the implementation I was also thinking about it, but a significant problem that I don't know how to handle is what bases actually we allow. Any Polygon? Any Ngon with the same N? How does a geometry look like, where a bottom geometry is a "banana shape", and the top is a "star shape"? It was easier to not care about these things.

Doing a little research I stumbled into Wolfram MathWorld and they have [Conical Frustum](https://mathworld.wolfram.com/ConicalFrustum.html) and [Pyramidal Frustum](https://mathworld.wolfram.com/PyramidalFrustum.html). Latter is  a special case for [Prismatoid](https://mathworld.wolfram.com/Prismatoid.html). Maybe the current `Frustum` should be named `ConicalFrustum`? And later the exact implementation of pyramidal frustums or prismatoids can be discussed.

**Comment by cserteGT3** [comment]
@juliohm I also did the requested changes.

All tests pass on my Windows 11 machine, under julia v1.9.3, not sure why those timeout on CI.

**Comment by juliohm** [comment]
I wonder if we could also try to support Ngon bases. We can merge this PR
as is and look at it in the future as well. 👍🏻

Tests sometimes fail on specific CI machines because our test suite has a
huge workload. I also think that some discretization algorithm has a hidden
bug that only occurs in our randomized tests sometimes. As soon as someone
reports a deterministic case we will be able to fix it.

Em ter., 12 de set. de 2023 06:08, cserteGT3 ***@***.***>
escreveu:

File: src/boundary.jl [hunk 6]
```diff
@@ -73,6 +73,10 @@ end
 
 boundary(::Torus) = nothing
 
+boundary(f::Frustum) = FrustumSurface(bottom(f), top(f))
+
+boundary(::FrustumSurface) = nothing
```
Comment (juliohm) [inline src/boundary.jl:76]: more to before torus

**Review by juliohm — APPROVED** [review]

**Pull request merged**